File: src/purra/delegation/coordinator.py
"""One-Run delegation coordination for isolated Agent executions."""

from __future__ import annotations

import asyncio
from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Protocol, runtime_checkable
from uuid import uuid4

from purra.cancellation import OperationCanceled, await_with_cancellation
from purra.contracts import (
    AgentDelegation,
    DelegationContextMode,
    DelegationAggregation,
    RunId,
    RuntimeOutcome,
)
from purra.json_values import freeze_json_mapping
from purra.normalization import required_text
from purra.operations import (
    AgentOperationController,
    OperationDisplay,
    OperationKind,
    OperationScope,
)
from purra.output import DelegationOutputEvent
from purra.output.processor import AgentOutputProcessor
from purra.ports import CancellationSignal, DelegationRepository


@dataclass(frozen=True, slots=True)
class DelegatedAgentRequest:
    """One Agent execution inside an existing Run."""

    run_id: RunId
    batch_id: str
    delegation_id: str
    agent_name: str
    agent_title: str
    agent_instruction: str
    objective: str
    input_payload: Mapping[str, Any] = field(default_factory=dict)
    context_mode: DelegationContextMode = DelegationContextMode.ISOLATED

    def __post_init__(self) -> None:
        for name, label in (
            ("run_id", "delegated execution run id"),
            ("batch_id", "delegation batch id"),
            ("delegation_id", "delegation id"),
            ("agent_name", "delegated Agent name"),
            ("agent_title", "delegated Agent title"),
            ("agent_instruction", "delegated Agent instruction"),
            ("objective", "delegation objective"),
        ):
            object.__setattr__(self, name, required_text(getattr(self, name), label))
        object.__setattr__(
            self,
            "input_payload",
            freeze_json_mapping(self.input_payload),
        )
        object.__setattr__(
            self,
            "context_mode",
            DelegationContextMode(self.context_mode),
        )


@dataclass(frozen=True, slots=True)
class DelegatedAgentResult:
    """Terminal result of a delegated execution, not a second Run result."""

    outcome: RuntimeOutcome
    content: str = ""
    error_code: str | None = None

    def __post_init__(self) -> None:
        outcome = RuntimeOutcome(self.outcome)
        error_code = str(self.error_code or "").strip() or None
        if outcome is RuntimeOutcome.COMPLETED and error_code is not None:
            raise ValueError("completed delegated execution cannot carry an error")
        if outcome is not RuntimeOutcome.COMPLETED and error_code is None:
            raise ValueError("unfinished delegated execution requires an error code")
        object.__setattr__(self, "outcome", outcome)
        object.__setattr__(self, "content", str(self.content or ""))
        object.__setattr__(self, "error_code", error_code)


@runtime_checkable
class DelegatedAgentExecutor(Protocol):
    async def execute(
        self,
        request: DelegatedAgentRequest,
        signal: CancellationSignal | None = None,
    ) -> DelegatedAgentResult: ...


class DelegationCoordinator:
    """Own one-shot delegated executions inside their existing Run."""

    def __init__(
        self,
        *,
        repository: DelegationRepository,
        executor: DelegatedAgentExecutor,
        output_processor: AgentOutputProcessor,
        operation_controller: AgentOperationController,
        max_parallel: int = 3,
    ) -> None:
        for method in (
            "create",
            "start",
            "complete",
            "fail",
            "cancel",
            "cancel_batch",
            "aggregate_batch",
        ):
            if not callable(getattr(repository, method, None)):
                raise TypeError(
                    "delegation coordinator requires a repository with "
                    f"{method}()"
                )
        if not isinstance(executor, DelegatedAgentExecutor):
            raise TypeError("delegation coordinator requires an Agent executor")
        self._repository = repository
        self._executor = executor
        self._output = output_processor
        self._operations = operation_controller
        self._capacity = asyncio.Semaphore(max(1, int(max_parallel)))
        self._active: dict[tuple[RunId, str], set[asyncio.Task[object]]] = {}

    async def create(
        self,
        *,
        run_id: RunId,
        batch_id: str,
        agent_name: str,
        agent_title: str,
        agent_instruction: str,
        objective: str,
        input_payload: Mapping[str, Any] | None = None,
        context_mode: DelegationContextMode = DelegationContextMode.ISOLATED,
        required: bool = True,
        priority: int = 0,
    ) -> AgentDelegation:
        delegation = await self._repository.create(
            run_id=run_id,
            batch_id=batch_id,
            agent_name=agent_name,
            agent_title=agent_title,
            agent_instruction=agent_instruction,
            objective=objective,
            input_payload=input_payload,
            context_mode=DelegationContextMode(context_mode),
            required=required,
            priority=priority,
        )
        await self._emit_status(delegation, "queued")
        return delegation

    async def execute_batch(
        self,
        delegations: Sequence[AgentDelegation],
        signal: CancellationSignal | None = None,
    ) -> tuple[DelegatedAgentResult, ...]:
        items = tuple(delegations)
        if not items:
            return ()
        run_id = items[0].run_id
        batch_id = items[0].batch_id
        if any(
            item.run_id != run_id or item.batch_id != batch_id
            for item in items
        ):
            raise ValueError("delegation batch must belong to one Run")

        tasks = tuple(
            asyncio.create_task(self._execute_one(item, signal))
            for item in items
        )
        key = (run_id, batch_id)
        active = self._active.setdefault(key, set())
        active.update(tasks)
        try:
            return tuple(await asyncio.gather(*tasks))
        finally:
            active.difference_update(tasks)
            if not active:
                self._active.pop(key, None)

    async def cancel_batch(self, run_id: RunId, batch_id: str) -> int:
        key = (
            required_text(run_id, "delegation Run id"),
            required_text(batch_id, "delegation batch id"),
        )
        canceled = await self._repository.cancel_batch(*key)
        tasks = tuple(self._active.get(key, ()))
        for task in tasks:
            task.cancel()
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)
        return canceled

    async def aggregate_batch(
        self,
        run_id: RunId,
        batch_id: str,
    ) -> DelegationAggregation:
        return await self._repository.aggregate_batch(run_id, batch_id)

    async def close(self) -> None:
        tasks = tuple(
            task
            for active in self._active.values()
            for task in active
        )
        for task in tasks:
            task.cancel()
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)
        self._active.clear()

    async def _execute_one(
        self,
        delegation: AgentDelegation,
        signal: CancellationSignal | None,
    ) -> DelegatedAgentResult:
        operation_id: str | None = None
        try:
            async with self._capacity:
                started = await self._repository.start(
                    delegation.id,
                    run_id=delegation.run_id,
                    batch_id=delegation.batch_id,
                )
                if started is None:
                    raise RuntimeError("delegation could not start")
                operation = await self._operations.start(
                    OperationKind.DELEGATION,
                    OperationScope(
                        run_id=delegation.run_id,
                        display=OperationDisplay(
                            label_key="agent.operation.delegation",
                            label_params={
                                "agentName": delegation.agent_name,
                                "delegationId": delegation.id,
                            },
                        ),
                    ),
                )
                operation_id = operation.operation_id
                await self._emit_status(started, "running")
                result = await await_with_cancellation(
                    self._executor.execute(
                        DelegatedAgentRequest(
                            run_id=delegation.run_id,
                            batch_id=delegation.batch_id,
                            delegation_id=delegation.id,
                            agent_name=delegation.agent_name,
                            agent_title=delegation.agent_title,
                            agent_instruction=delegation.agent_instruction,
                            objective=delegation.objective,
                            input_payload=delegation.input_payload,
                            context_mode=delegation.context_mode,
                        ),
                        signal,
                    ),
                    signal,
                )
                if not isinstance(result, DelegatedAgentResult):
                    raise TypeError(
                        "delegated Agent executor returned an invalid result"
                    )
                await self._settle(delegation, result, operation_id)
                return result
        except OperationCanceled:
            await self._cancel(delegation, operation_id)
            return DelegatedAgentResult(
                outcome=RuntimeOutcome.CANCELED,
                error_code="delegation_canceled",
            )
        except asyncio.CancelledError:
            await self._cancel(delegation, operation_id)
            raise
        except Exception as error:
            code = str(getattr(error, "code", "") or type(error).__name__)
            await self._repository.fail(
                delegation.id,
                run_id=delegation.run_id,
                batch_id=delegation.batch_id,
                error=code,
            )
            if operation_id is not None:
                await self._operations.fail(operation_id, code)
            await self._emit_status(
                delegation,
                "failed",
                error_code=code,
            )
            return DelegatedAgentResult(
                outcome=RuntimeOutcome.FAILED,
                error_code=code,
            )

    async def _cancel(
        self,
        delegation: AgentDelegation,
        operation_id: str | None,
    ) -> None:
        await self._repository.cancel(
            delegation.id,
            run_id=delegation.run_id,
            batch_id=delegation.batch_id,
            reason="delegation_canceled",
        )
        if operation_id is not None:
            await self._operations.cancel(
                operation_id,
                "delegation_canceled",
            )
        await self._emit_status(
            delegation,
            "canceled",
            error_code="delegation_canceled",
        )

    async def _settle(
        self,
        delegation: AgentDelegation,
        result: DelegatedAgentResult,
        operation_id: str,
    ) -> None:
        if result.outcome is RuntimeOutcome.COMPLETED:
            settled = await self._repository.complete(
                delegation.id,
                run_id=delegation.run_id,
                batch_id=delegation.batch_id,
                result_summary=result.content,
            )
            if not settled:
                raise RuntimeError("delegation result could not be persisted")
            await self._operations.succeed(operation_id)
            await self._emit_status(delegation, "done")
            return
        if result.outcome is RuntimeOutcome.CANCELED:
            await self._repository.cancel(
                delegation.id,
                run_id=delegation.run_id,
                batch_id=delegation.batch_id,
                reason=result.error_code or "delegation_canceled",
            )
            await self._operations.cancel(
                operation_id,
                result.error_code or "delegation_canceled",
            )
            await self._emit_status(
                delegation,
                "canceled",
                error_code=result.error_code,
            )
            return
        await self._repository.fail(
            delegation.id,
            run_id=delegation.run_id,
            batch_id=delegation.batch_id,
            error=result.error_code or "delegation_failed",
        )
        await self._operations.fail(
            operation_id,
            result.error_code or "delegation_failed",
        )
        await self._emit_status(
            delegation,
            "failed",
            error_code=result.error_code,
        )

    async def _emit_status(
        self,
        delegation: AgentDelegation,
        status: str,
        *,
        error_code: str | None = None,
    ) -> None:
        await self._output.accept_delegation_event(DelegationOutputEvent(
            event_id=f"delegation-{uuid4().hex}",
            run_id=delegation.run_id,
            batch_id=delegation.batch_id,
            delegation_id=delegation.id,
            status=status,
            agent_name=delegation.agent_name,
            agent_title=delegation.agent_title,
            objective=delegation.objective,
            error_code=error_code,
            occurred_at=datetime.now(timezone.utc),
        ))


__all__ = [
    "DelegatedAgentExecutor",
    "DelegatedAgentRequest",
    "DelegatedAgentResult",
    "DelegationContextMode",
    "DelegationCoordinator",
]

File: typescript/src/delegation/coordinator.ts
import { copyJsonValue } from "../model/validation.js";
import { AgentCanceledError, AgentError } from "../shared/errors.js";
import { stableFingerprint } from "../shared/fingerprint.js";
import { DelegationPolicy } from "./policy.js";
import type {
  AgentDelegation,
  DelegatedAgentExecutor,
  DelegatedAgentResult,
  DelegationAggregation,
  DelegationDefinition,
  DelegationEventSink,
  DelegationLifecycleEvent,
  DelegationRepository,
} from "./types.js";

interface ActiveCall {
  readonly digest: string;
  readonly promise: Promise<DelegationAggregation>;
}

export class DelegationCoordinator {
  readonly #repository: DelegationRepository;
  readonly #executor: DelegatedAgentExecutor;
  readonly #policy: DelegationPolicy;
  readonly #capacity: Semaphore;
  readonly #sinks = new Map<string, DelegationEventSink>();
  readonly #activeCalls = new Map<string, ActiveCall>();
  readonly #controllers = new Map<string, Set<AbortController>>();

  public constructor(options: {
    readonly repository: DelegationRepository;
    readonly executor: DelegatedAgentExecutor;
    readonly policy?: DelegationPolicy;
  }) {
    if (typeof options.repository?.createBatch !== "function") {
      throw new TypeError("Delegation coordinator requires a repository");
    }
    if (typeof options.executor?.execute !== "function") {
      throw new TypeError("Delegation coordinator requires an executor");
    }
    this.#repository = options.repository;
    this.#executor = options.executor;
    this.#policy = options.policy ?? new DelegationPolicy();
    this.#capacity = new Semaphore(this.#policy.snapshot().maxParallel);
  }

  public bindRun(runId: string, sink: DelegationEventSink): void {
    const id = requiredText(runId, "delegation Run id");
    if (this.#sinks.has(id)) throw new AgentError("delegation_root_run_already_bound", "Root Run is already bound");
    this.#sinks.set(id, sink);
  }

  public releaseRun(runId: string): void {
    this.#sinks.delete(runId);
  }

  public async executeCall(input: {
    readonly runId: string;
    readonly idempotencyKey: string;
    readonly delegations: unknown;
    readonly enabledTools?: readonly string[];
    readonly signal?: AbortSignal;
  }): Promise<DelegationAggregation> {
    const runId = requiredText(input.runId, "delegation Run id");
    const idempotencyKey = requiredText(input.idempotencyKey, "delegation idempotency key");
    if (!this.#sinks.has(runId)) {
      throw new AgentError("delegation_root_run_not_active", "Delegation requires an active Root Run");
    }
    const definitions = this.#policy.validate(input.delegations);
    const digest = await stableFingerprint(copyJsonValue(definitions));
    const key = `${runId}\u0000${idempotencyKey}`;
    const active = this.#activeCalls.get(key);
    if (active !== undefined) {
      if (active.digest !== digest) {
        throw new AgentError(
          "delegation_idempotency_conflict",
          "Delegation idempotency key was reused with different input",
        );
      }
      return active.promise;
    }
    const promise = this.#executeCall({
      runId,
      idempotencyKey,
      definitions,
      ...(input.enabledTools === undefined ? {} : { enabledTools: Object.freeze([...input.enabledTools]) }),
      ...(input.signal === undefined ? {} : { signal: input.signal }),
    });
    this.#activeCalls.set(key, { digest, promise });
    try {
      return await promise;
    } finally {
      this.#activeCalls.delete(key);
    }
  }

  public async cancelBatch(runId: string, batchId: string): Promise<number> {
    const id = requiredText(runId, "delegation Run id");
    const batch = requiredText(batchId, "delegation batch id");
    const rows = (await this.#repository.listForRun(id)).filter((row) => row.batchId === batch);
    const canceled = await this.#repository.cancelBatch(id, batch);
    for (const controller of this.#controllers.get(batchKey(id, batch)) ?? []) controller.abort();
    const settled = new Map((await this.#repository.listForRun(id))
      .filter((row) => row.batchId === batch)
      .map((row) => [row.id, row]));
    for (const row of rows) {
      if (
        (row.status === "queued" || row.status === "running")
        && settled.get(row.id)?.status === "canceled"
      ) await this.#emit(row, "canceled", "delegation_canceled");
    }
    return canceled;
  }

  public async close(): Promise<void> {
    for (const controllers of this.#controllers.values()) {
      for (const controller of controllers) controller.abort();
    }
    await Promise.allSettled([...this.#activeCalls.values()].map((call) => call.promise));
    this.#controllers.clear();
    this.#activeCalls.clear();
    this.#sinks.clear();
  }

  async #executeCall(input: {
    readonly runId: string;
    readonly idempotencyKey: string;
    readonly definitions: readonly DelegationDefinition[];
    readonly enabledTools?: readonly string[];
    readonly signal?: AbortSignal;
  }): Promise<DelegationAggregation> {
    const batchId = `delegation-batch:${input.idempotencyKey}`;
    const receipt = await this.#repository.createBatch({
      runId: input.runId,
      batchId,
      idempotencyKey: input.idempotencyKey,
      delegations: input.definitions,
    });
    if (!receipt.replayed) {
      for (const row of receipt.delegations) await this.#emit(row, "queued");
    }
    const executable: AgentDelegation[] = [];
    for (const row of receipt.delegations) {
      if (row.status === "queued") executable.push(row);
      if (row.status === "running") {
        const changed = await this.#repository.fail(
          row.id,
          row.runId,
          row.batchId,
          "delegation_interrupted",
        );
        if (changed) await this.#emit(row, "failed", "delegation_interrupted");
      }
    }
    await Promise.all(executable.map((row) => this.#executeOne(
      row,
      input.enabledTools,
      input.signal,
    )));
    return this.#repository.aggregateBatch(input.runId, batchId);
  }

  async #executeOne(
    delegation: AgentDelegation,
    enabledTools: readonly string[] | undefined,
    parentSignal: AbortSignal | undefined,
  ): Promise<DelegatedAgentResult> {
    const release = await this.#capacity.acquire(parentSignal);
    const controller = new AbortController();
    const key = batchKey(delegation.runId, delegation.batchId);
    const controllers = this.#controllers.get(key) ?? new Set<AbortController>();
    controllers.add(controller);
    this.#controllers.set(key, controllers);
    const signal = parentSignal === undefined
      ? controller.signal
      : AbortSignal.any([parentSignal, controller.signal]);
    try {
      const started = await this.#repository.start(delegation.id, delegation.runId, delegation.batchId);
      if (started === undefined) {
        return Object.freeze({ outcome: "failed", errorCode: "delegation_could_not_start" });
      }
      await this.#emit(started, "running");
      const result = validateResult(await this.#executor.execute(Object.freeze({
        runId: started.runId,
        batchId: started.batchId,
        delegationId: started.id,
        agentName: started.agentName,
        agentTitle: started.title,
        agentInstruction: started.instruction,
        objective: started.objective,
        input: started.input,
        contextMode: "isolated",
        ...(enabledTools === undefined ? {} : { enabledTools }),
      }), signal));
      if (result.outcome === "completed") {
        if (!await this.#repository.complete(
          started.id,
          started.runId,
          started.batchId,
          result.content ?? "",
        )) {
          throw new AgentError("delegation_settlement_failed", "Delegation result could not be persisted");
        }
        await this.#emit(started, "done");
        return result;
      }
      const status = result.outcome === "canceled" ? "canceled" : "failed";
      const changed = status === "canceled"
        ? await this.#repository.cancel(started.id, started.runId, started.batchId, result.errorCode!)
        : await this.#repository.fail(started.id, started.runId, started.batchId, result.errorCode!);
      if (changed) await this.#emit(started, status, result.errorCode);
      return result;
    } catch (error) {
      const canceled = signal.aborted || error instanceof AgentCanceledError;
      const errorCode = canceled ? "delegation_canceled" : code(error);
      const changed = canceled
        ? await this.#repository.cancel(delegation.id, delegation.runId, delegation.batchId, errorCode)
        : await this.#repository.fail(delegation.id, delegation.runId, delegation.batchId, errorCode);
      if (changed) await this.#emit(delegation, canceled ? "canceled" : "failed", errorCode);
      if (!canceled && !changed) throw error;
      return Object.freeze({ outcome: canceled ? "canceled" : "failed", errorCode });
    } finally {
      controllers.delete(controller);
      if (controllers.size === 0) this.#controllers.delete(key);
      release();
    }
  }

  async #emit(row: AgentDelegation, status: DelegationLifecycleEvent["status"], errorCode?: string): Promise<void> {
    const sink = this.#sinks.get(row.runId);
    if (sink === undefined) throw new AgentError("delegation_root_run_not_active", "Root Run is not active");
    await sink(Object.freeze({
      runId: row.runId,
      batchId: row.batchId,
      delegationId: row.id,
      agentName: row.agentName,
      agentTitle: row.title,
      status,
      ...(errorCode === undefined ? {} : { errorCode }),
    }));
  }
}

class Semaphore {
  readonly #limit: number;
  #active = 0;
  readonly #waiting: (() => void)[] = [];

  public constructor(limit: number) {
    this.#limit = limit;
  }

  public async acquire(signal?: AbortSignal): Promise<() => void> {
    if (signal?.aborted) throw new AgentCanceledError();
    if (this.#active >= this.#limit) {
      await new Promise<void>((resolve, reject) => {
        const cancel = (): void => {
          const index = this.#waiting.indexOf(resume);
          if (index >= 0) this.#waiting.splice(index, 1);
          reject(new AgentCanceledError());
        };
        const resume = (): void => {
          signal?.removeEventListener("abort", cancel);
          resolve();
        };
        this.#waiting.push(resume);
        signal?.addEventListener("abort", cancel, { once: true });
      });
    }
    this.#active += 1;
    let released = false;
    return () => {
      if (released) return;
      released = true;
      this.#active -= 1;
      this.#waiting.shift()?.();
    };
  }
}

function validateResult(value: DelegatedAgentResult): DelegatedAgentResult {
  if (value === null || typeof value !== "object") throw new TypeError("Invalid delegated Agent result");
  if (value.outcome !== "completed" && value.outcome !== "failed" && value.outcome !== "canceled") {
    throw new TypeError("Invalid delegated Agent outcome");
  }
  const errorCode = value.errorCode === undefined ? undefined : requiredText(value.errorCode, "delegation error code");
  if (value.outcome === "completed" && errorCode !== undefined) {
    throw new TypeError("Completed delegation cannot carry an error");
  }
  if (value.outcome !== "completed" && errorCode === undefined) {
    throw new TypeError("Unfinished delegation requires an error code");
  }
  return Object.freeze({
    outcome: value.outcome,
    ...(value.content === undefined ? {} : { content: copyJsonValue(value.content) }),
    ...(errorCode === undefined ? {} : { errorCode }),
  });
}

function batchKey(runId: string, batchId: string): string {
  return `${runId}\u0000${batchId}`;
}

function code(error: unknown): string {
  return error instanceof AgentError ? error.code : "delegation_failed";
}

function requiredText(value: unknown, label: string): string {
  const text = typeof value === "string" ? value.trim() : "";
  if (text === "") throw new TypeError(`${label} must be non-empty text`);
  return text;
}
